source/common/upstream/upstream_impl.cc

```diff
@@ -607,6 +607,31 @@ void PrioritySetImpl::BatchUpdateScope::updateHosts(
                       hosts_removed, overprovisioning_factor);
 }
 
+void MainPrioritySetImpl::updateMutableAllHostMap(const HostVector& hosts_added,
+                                                  const HostVector& hosts_removed) {
+  if (hosts_added.empty() && hosts_removed.empty()) {
+    // No new hosts have been added and no old hosts have been removed.
+    return;
+  }
+
+  // Since read_only_all_host_map_ may be shared by multiple threads, when the host set changes, we
+  // cannot directly modify read_only_all_host_map_.
+  if (mutable_all_host_map_ == nullptr) {
+    // Copy old read only host map to mutable host map.
+    mutable_all_host_map_ = read_only_all_host_map_ != nullptr
+                                ? std::make_shared<HostMap>(*read_only_all_host_map_)
+                                : std::make_shared<HostMap>();
```

Comment:
Would this be prohibitive for large clusters? Even if just one host changes, we end up doing an O(n) copy of this map. 

Speaking of large clusters, since this would add both space and computational overhead to large clusters, should we make this configurable or only enabled when actually used in some way?

Reply:
I think we are already paying this copy cost for all the other data structures, right? (Not that it's good to add yet another one.). One option for this particular structure would be to use R/W locks and then we wouldn't need to actually do any copies on it.

Reply:
I guess we do pay this when processing the EDS update as a whole, so we wouldn't make it that much worse. 

I was originally thinking in terms of the update callbacks, most (all?) of which operate on the per priority level, so a change in P0 won't trigger updates to P1, P2, ...

Reply:
Yeah, good point re: per-priority, though I'm not sure how much that matters in practice for most users. I think though that as long as we can replace the EDS private one with this one it's probably a wash?

Reply:
Yep if this lets us replace the `all_hosts_` map then I think we're good

Reply:
> Yep if this lets us replace the `all_hosts_` map then I think we're good

Yep. This will replace the `all_hosts_` map in EDS. This may still bring additional overhead, because each priority may be updated once, causing the callback to be executed multiple times. In the original EDS, because `batchUpdate` was used, `all_hosts_` would only be copied once. But this problem can be alleviated by setting `merge_timeout`.

Another method is to use an additional `memberUpdateCallback` to synchronize `read_only_all_host_map_` to the worker threads instead of directly reusing `postThreadLocalClusterUpdate`. This ensures that when `batchUpdate` is used, there is only one map copy at most. I think this is also a point that can be discussed.

Reply:
> I think we are already paying this copy cost for all the other data structures, right? (Not that it's good to add yet another one.). One option for this particular structure would be to use R/W locks and then we wouldn't need to actually do any copies on it.

I have considered using R/W locks, but I don't want to add the overhead of acquiring R locks once during each request for the occasional membership update.